system-tests/tests/src/test/java/org/eclipse/dataspaceconnector/system/tests/utils/FileTransferSimulationUtils.java

```diff
@@ -0,0 +1,149 @@
+ */
+public abstract class FileTransferSimulationUtils {
+
+    private FileTransferSimulationUtils() {
+    }
+
+    public static final String DESCRIPTION = "[Contract negotiation and file transfer]";
+    public static final String OFFER_FILE = "contractoffer.json";
+
+    public static final String PROVIDER_ASSET_NAME = "test-document";
+
+    private static final String CONNECTOR_ADDRESS_PARAM = "connectorAddress";
+    private static final String DESTINATION_PARAM = "destination";
+    private static final String CONTRACT_ID_PARAM = "contractId";
+    private static final String API_KEY_HEADER = "X-Api-Key";
+
+    private static final Faker FAKER = new Faker();
+
+    /**
+     * Gatling chain for performing contract negotiation and file transfer.
+     *
+     * @param providerUrl     URL for the Provider API, as accessed from the Consumer runtime.
+     * @param destinationPath File copy destination path. If it includes the character sequence {@code %s}, that sequence is replaced with a random string in each iteration.
+     * @param apiKey          Consumer runtime API Key.
+     */
+    public static ChainBuilder contractNegotiationAndFileTransfer(String providerUrl, String destinationPath, String apiKey) {
+        String connectorAddress = format("%s/api/ids/multipart", providerUrl);
+        String body;
+        try {
+            InputStream offerFileData = Thread.currentThread().getContextClassLoader().getResourceAsStream(OFFER_FILE);
+            body = new String(Objects.requireNonNull(offerFileData, OFFER_FILE + " not found").readAllBytes());
+        } catch (IOException e) {
+            throw new EdcException(e);
+        }
+        return group("Contract negotiation")
+                .on(exec(
+                        // Initiate a contract negotiation
+                        http("Contract negotiation")
+                                .post("/api/negotiation")
+                                .body(StringBody(body))
+                                .header(CONTENT_TYPE, "application/json")
+                                .queryParam(CONNECTOR_ADDRESS_PARAM, connectorAddress)
+                                .check(status().is(SC_OK))
+                                .check(bodyString()
+                                        .notNull()
+                                        // UUID is returned to get the contract agreement negotiated between provider and consumer.
+
+                                        .saveAs("contractNegotiationRequestId"))
+                ))
+                // Call ContractNegotiation status endpoint every second, till it gives a CONFIRMED state.
+                .exec(session -> session.set("status", -1))
+                .group("Wait for agreement").on(
+                        doWhileDuring(session -> session.getString("contractAgreementId") == null,
+                                Duration.ofSeconds(30))
+                                .on(exec(http("Get status")
```

Comment:
For better readability how about if we can extract all _exec(_ function body to separate methods ?

Reply:
Thank you! I hadn't realized the DSL is so powerful you can rewrite .group() chains into .exec(group()) which suit themselves for refactoring :)